### PR TITLE
Bunch of fixes for CA1859

### PR DESCRIPTION
**Pull request by geeknoid**
* When using the ?? operator, the nullable annotation for the left-hand operand is now erased. This prevents the analyzer suggesting to use a replacement nullable type rather than its non-nullable variation. #6413 

* We no longer suggest to upgrade the type of a local/field/parameter if the symbol is being used to invoke a method that is an explicit implementation of an interface method. If the user would upgrade the type, the call to that method would no longer work. #6416 

* Ensure that we never recommend upgrading the signature of a method that implements an interface method.

* Don't recommend a method to be upgraded if the method is an implementation of a partial method definition. This is because there might be different implementations of the method with conflicting behavior. Note that if you use #if constructs, the diagnostic may still make recommendations that would break your code since the analyzer only knows about the selected #if block. #6414 

* Remove a field-specific optimization that was designed to speed up the analyzer, since it actually broke analysis of fields, yielding bogus analysis results. #6411

### CODE REVIEW

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs:305]: Can you add a test for the following?

```csharp
                    internal class C
                    {
                        private MemoryStream? _stream;

                        private Stream? {|#0:GetStream|}()
                        {
                            return _stream ?? Create();
                        }

                        private MemoryStream? Create() => new MemoryStream();
                    }
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs
Comment (Youssef1313) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs:369]: Can `values` be still empty here? (either valid scenario or error scenario)

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs
Comment (geeknoid) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs:274]: Done.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs
Comment (geeknoid) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeAnalyzer.Collector.cs]: Good point, I added a guard for that.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs
Comment (Youssef1313) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs:312]: I think suggested replacement should be `System.IO.MemoryStream?` for this scenario. (ie, we should preserve the original nullable annotation). Thoughts?

Comment (geeknoid) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs:312]: I was actually wondering about the opposite situation. Should I make a point of suggesting upgrades in nullability state in general? So:

```csharp
 object? Foo() => return new object();
```

would suggest upgrading the return value to `object`

I do have a separate analyzer that I want to bring in to help eliminate spurious uses of ?, ??, and !. But that other analyzer is focused on expression-level analysis and wouldn't tackle this particular case.

Comment (buyaa-n) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs:312]: I support @Youssef1313 's point `preserve the original nullable annotation`, what you think @stephentoub?

> I was actually wondering about the opposite situation. Should I make a point of suggesting upgrades in nullability state in general? 

Such nullability upgrades probably should be added to the Roslyn nullability analysis instead.

**Review by buyaa-n — APPROVED** [review]
Overall LGTM, thank you! Will wait for the open question resolved before merge

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs
Comment (geeknoid) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/UseConcreteTypeTests.cs]: OK done.

**Comment by Youssef1313** [comment]
#6411 #6413 #6414 #6416 @geeknoid Are all these issues fixed and can be closed?

**Comment by geeknoid** [comment]
@Youssef1313  Yes, all these issues were fixed in my previosu PR.

**Comment by Youssef1313** [comment]
Thanks @geeknoid. I went ahead and closed the four issues.

**Pull request merged**
